dotnet-authserver/src/TeacherIdentity.AuthServer/Journeys/TrnTokenHelper.cs

```diff
@@ -0,0 +1,171 @@
+        _clock = clock;
+        _dqtApiClient = dqtApiClient;
+        _configuration = configuration;
+        _userSearchService = userSearchService;
+    }
+
+    public async Task<EnhancedTrnToken?> GetValidTrnToken(OpenIddictRequest request)
+    {
+        var requestedTrnToken = request["trn_token"];
+
+        if (_configuration.GetValue("RegisterWithTrnTokenEnabled", false) && requestedTrnToken.HasValue)
+        {
+            return await GetValidTrnTokenModel(requestedTrnToken.Value.Value as string);
+        }
+
+        return null;
+    }
+
+    public async Task InitializeAuthenticationStateWithToken(
+        User? signedInUser,
+        AuthenticationState authenticationState,
+        EnhancedTrnToken trnToken,
+        HttpContext httpContext)
+    {
+        if (signedInUser is not null)
+        {
+            if (signedInUser.EmailAddress == trnToken.Email)
+            {
+                authenticationState.OnSignedInUserProvided(signedInUser);
+                if (signedInUser.Trn is null)
+                {
+                    authenticationState.OnTrnTokenProvided(trnToken);
+                }
+            }
+            else
+            {
+                authenticationState.OnTrnTokenProvided(trnToken);
+            }
+        }
+        else
+        {
+            var existingValidUser = await _dbContext.Users.SingleOrDefaultAsync(u => u.EmailAddress == trnToken.Email);
+
+            if (existingValidUser is not null)
+            {
+                if (existingValidUser.Trn is not null && existingValidUser.Trn != trnToken.Trn)
+                {
+                    // If we find a matching email in our records but TRN no match, ignore the token
+                    return;
+                }
+
+                authenticationState.OnSignedInUserProvided(existingValidUser);
+                if (existingValidUser.Trn is null)
+                {
+                    authenticationState.OnTrnTokenProvided(trnToken);
+                }
+                await authenticationState.SignIn(httpContext);
+            }
+            else
+            {
+                var users = await _userSearchService.FindUsers(trnToken.FirstName, trnToken.LastName, trnToken.DateOfBirth);
+                authenticationState.OnExistingAccountSearch(users.Length == 0 ? null : users[0]);
+                authenticationState.OnTrnTokenProvided(trnToken);
+            }
+        }
+    }
+
+    public async Task<User> ApplyTrnTokenToUser(Guid? userId, string trnTokenValue)
+    {
+        var user = await _dbContext.Users.SingleAsync(u => u.UserId == userId);
+        var trnToken = await _dbContext.TrnTokens.SingleAsync(t => t.TrnToken == trnTokenValue);
+
+        if (user.Trn is null)
+        {
+            user = await UpdateUserTrn(user, trnToken.Trn);
+        }
+
+        if (user.Trn == trnToken.Trn)
+        {
+            await InvalidateTrnToken(trnToken.TrnToken, user.UserId);
+        }
+
+        return user;
```

Comment:
This is the main reason why the ACs on a couple of tests are don't match the ticket. If we were to implement as requested I don't think this function could exist, we would have to invalidate tokens on a case by case basis (and probably add some parameters to determine the current edge case we are trying to handle)